FW-863. 6LoWPAN fragmentation module

TimothyClaeys opened the pull request:
This pull request implements 6LoWPAN fragmentation.
- PR is currently configured to support up to 1280 bytes of IPv6 data
- implements fast-fragment forwarding on the intermediate hops to reduce latency
- reassembly timeout mechanism (orphan fragments get dropped after 60s)
- this PR update files/code across the project :
   --> definition of large packets in opendefs.h and openqueue.h/.c
   --> fully allocate SRAM memory (linker scripts for the openmotes)
   --> updates to the packet_functions to support large packets ( > 127 bytes)
- this PR works in combination with the openvisualizer code which implements fragmentation and reassembly for outward and incoming traffic.

Commit:
FW-863. Commit with partial implementation of the 6LoWPAN fragmentation feature.

(1) Copy address information from original message to fragments

(2) Bug fixes and further integration of 6LoWPAN fragmentation.
    - fix checksum calculation of UPD and TCP (in case message length > 128)
    - allow uecho to allocate OpenQueueBigPackets
    - remove some useless code from openbridge
    - update packetfunctions to handle big packets (> 127 bytes)
    - update the uecho.py script to be more interactive

Commit:
FW-863. Add support for simulation mode

TimothyClaeys committed:
FW-863. Reformatted iphc.c

TimothyClaeys committed:
FW-863. Delete redundant code from IPv6 header parsing function

TimothyClaeys committed:
FW-863. Correctly format forwarding.c

TimothyClaeys committed:
FW-863. Add support for source routing of fragments

 (1) Bug fixes for upward routing.

 (2) Fixed a bug in fragmentation layer.

 (3) Update definitions opendefs.h

 (4) Support for out-of-order fragment fast-forwarding

TimothyClaeys committed:
FW-863. Correctly formatted packetfunctions.c/.h

TimothyClaeys committed:
FW-863. Formatting uecho.c

TimothyClaeys committed:
FW-863. Add a reassembly timeout mechanism

+ support for large ICMPv6 echo response

TimothyClaeys committed:
FW-863. Fixes issue when forwarding out-of-order fragments.

TimothyClaeys committed:
FW-863. Fixes the fragment buffer and VRB cleanup functions

When a fragment is received, a node schedules a 60 timeout either for the reassembly buffer or fast-forwarding VRB. If not all bytes are received within this 60s interval, all fragments are dropped and the buffers are liberated.

TimothyClaeys committed:
FW-863. Refactor openbridge.c/.h

TimothyClaeys committed:
Merge branch 'develop' into develop_FW-863

TimothyClaeys committed:
FW-863. Fix segmentation fault when L2 security is active

- When L2 security is active, the fragmentation payload size must be reduced to make place for the auxiliary security header and the authentication tag

TimothyClaeys committed:
FW-863. Allocate all the SRAM memory

+ Tell linker that there is 32kB of SRAM instead of 16kB.

TimothyClaeys committed:
FW-863. Fix compile errors macpong project and samr21_xpro

TimothyClaeys committed:
FW-863. Bug fixes and formatting

- implicit declarations
- out-of-bounds array
- clang-tidy remarks

TimothyClaeys committed:
FW-863. Additional log for fragmentation

TimothyClaeys committed:
FW-863. Cleanup fragments if buffer allocation fails

changtengfei commented on openstack/03a-IPHC/frag.c:
```diff
@@ -0,0 +1,756 @@
+        }
+
+        // send all the fragments with the current datagram tag
+        for (int i = 0; i < FRAGMENT_BUFFER_SIZE; i++) {
```
need to replace int with 'stdint.h definition'

changtengfei commented on openstack/03a-IPHC/frag.c:
```diff
@@ -0,0 +1,756 @@
+                fragment_length = remaining_bytes;
+
+            // find a new spot in the fragmentation buffer  
+            for (int i = 0; i < FRAGMENT_BUFFER_SIZE; i++) {
```
need to replace int with 'stdint.h definition'.

I prefer to define i in advance, rather than define it in side the for loop, like here. 
You may reuse the 'i' latter.

changtengfei commented on openstack/03a-IPHC/frag.c:
```diff
@@ -0,0 +1,756 @@
+        upward_relay = FALSE;
+
+        // delete sent fragment from fragment buffer
+        int i;
```
Usually, I define all various at the beginning of the function. Merely a C99 style.
Bit, this is trivle.

changtengfei commented on openstack/03a-IPHC/frag.c:
```diff
@@ -0,0 +1,756 @@
+    uint16_t received_bytes;
+
+    bool do_reassemble = FALSE;
+    bool has_timer = FALSE;
```
I would prefer separate this code by defining it first than assign values.

changtengfei commented on openstack/03a-IPHC/frag.c:
```diff
@@ -0,0 +1,756 @@
+    }
+
+    // we find buffer space for a new fragment (new datagram_tag)
+    int i;
```
try to keep it consistent.

changtengfei commented on openstack/03a-IPHC/frag.c:
```diff
@@ -0,0 +1,756 @@
+    total_wanted_bytes = size;
+    received_bytes = dropped_srh_len = count = 0;
+
+    for (int j = 0; j < FRAGMENT_BUFFER_SIZE; j++) {
```
why not use i here?

changtengfei reviewed (commented):
@TimothyClaeys the code looks great!  Just some trival comments which you can resolve it easily.

TimothyClaeys committed:
FW-863. Formatting code to make it compatible with GCC-4.8

TimothyClaeys committed:
FW-863. Reformat code openqueue.c

Fix indentation and spacing in the file.

TimothyClaeys commented:
Commits address remarks @changtengfei and fix issues when compiling with GCC 4.8.

TimothyClaeys committed:
FW-863. Bug fix related to "small" big packets

When a module request an OpenQueueBigEntry_t packet, but the total size of the packet did not exceed the fragmentation limit, the packet was never scheduled for transmission by the MAC layer.
